studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -19,14 +19,133 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson; 
+
+import com.google.sps.models.Student;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+    // Value to filter students by while Querying.
+    private String subject;
+
+/********************
+Handles GET requests
+********************/
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    // TODO Integrate UserService
+    final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    // Query Datastore
+    Filter subject_filter = new FilterPredicate("subject", FilterOperator.EQUAL, this.subject);
+    Query query = new Query("Student");
+    query.setFilter(subject_filter);
+    query.addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList <Student> students = new ArrayList<>();
+
+    // Scan results and add the in the students ArrayList
+    for (Entity entity: results.asIterable()){
+        Student current_student = createStudentFromEntity(entity);
+        students.add(current_student);
+    }
+
+    // Return a JSON response of the results
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJson(students));
+  }
```

Comment:
I'm not sure how this is going to work. From our design meeting, I thought the plan was to create a single POST request that searches for a match and sends the person's info to the datastore, but here it looks like that is split across a GET and POST which I believe will create a race condition.

One solution might be to move the functionality of the GET request to a private function and call it from inside the POST request.

Reply:
Thanks for the feedback. I will address this.

@mpasetes17 @sgh005 
Quick question for clarification: Among students who match  the ```subject``` and ```school``` criteria requested, are we only sending back 

- A single student. 
- All students.
- Or maybe a Top N number of students based on ```timestamp```.

For now, it is returning **all**. 

Ben

Reply:
I think returning all is fine. If it becomes an issue, I don't think it will take much to add a limit.